[FSDP] Add FlatParameter to track the information of a flat parameter

fegin opened the pull request:
Summary: Implement FlatParameter to track the information of a flat parameter, including the sharding information.

Test Plan: CI

Differential Revision: D32432503



cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

zhaojuanmao commented on torch/distributed/_fsdp/flatten_params_wrapper.py:
```diff
@@ -178,21 +305,13 @@ def _get_param_views(
+        assert self.flat_param is not None
+        return self.flat_param.get_param_views(external_data)
 
     def metadata(self) -> Tuple[List[str], List[torch.Size], List[int]]:
```
this can be removed? as flat_param already has this?

zhaojuanmao commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -194,8 +194,10 @@ def __init__(
         for n, p in self.named_parameters():
-            if not hasattr(p, "_is_sharded"):
-                raise RuntimeError(f"found unsharded parameter: {n} ; {p.size()}")
+            if not isinstance(p, FlatParameter):
```
let us still check 'if not hasattr(p, "_is_sharded")', which is more strict to require parameters are sharded

fegin replied:
Why this is more strict? `hasattr(p, "_is_sharded")` only test whether the `_is_sharded` member variable exists or not. `_is_sharded` can be False if `world_size` is 1 (or some other cases in the future). IMHO,  testing if the parameter is `FlatParameter` or not is more appropriate as in the below code we use it as it is a `FlatParameter` (e.g., assuming it has some APIs/member variables like `shard_metadata`).

zhaojuanmao commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -177,7 +177,7 @@ def __init__(
         params = []
         for param_name, param in module.named_parameters():
-            if not hasattr(param, "_is_sharded"):
+            if not isinstance(param, FlatParameter):
```
let us still check 'if not hasattr(p, "_is_sharded")', which is more strict to require parameters are sharded

fegin replied:
comment above, let's discuss there.

zhaojuanmao commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -283,11 +285,8 @@ def _shard_parameters(self) -> None:
-            assert not hasattr(
-                p, "_is_sharded"
-            ), "Param should have not been sharded yet."
+            assert not p._is_sharded, "Param should have not been sharded yet."
```
same as above

zhaojuanmao commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -864,6 +868,8 @@ def update_p_data(output_tensor: torch.Tensor) -> None:
             # Trim any padding and reshape to match original size.
             p.data = p.data[: p._orig_size.numel()].view(p._orig_size)  # type: ignore[attr-defined]
 
+        self._lazy_init()
```
why is it called here?

fegin replied:
There was a parameters-uninitialized corner case when testing.  The case seems to be fixed or not exist anymore. Will remove this.

zhaojuanmao reviewed (commented):
Thanks for adding this! overall looks good, please see some nits and minor comments

rohan-varma commented on test/distributed/fsdp/test_flatten_params_wrapper.py:
```diff
@@ -170,6 +170,111 @@ def test_shared_params_pnorm_after_step(self):
+        def _test(kwargs, expected, exception=None):
+            flat_p._is_sharded = True
+            if exception is not None:
+                with self.assertRaises(exception):
```
nit: `self.assertRaisesRegex` is usually better.

rohan-varma reviewed (commented):
Thanks for adding this! Have a couple of comments inline.

fegin commented on torch/distributed/_fsdp/flatten_params_wrapper.py:
```diff
@@ -6,19 +6,159 @@
+        self._param_numels = [p.numel() for p in params]
+        assert self.numel() <= sum(
+            self._param_numels
+        ), f"Something wrong with __new__ method, {self.numel()} vs. {sum(self._param_numels)}"
```
This is `__init__`, but the data is initialized in `__new__`. I changed the message to make it more informative.

fegin commented on torch/distributed/_fsdp/flatten_params_wrapper.py:
```diff
@@ -6,19 +6,159 @@
+        )  # type: ignore[call-arg]
+
+    def __init__(self, params: Sequence[nn.Parameter], requires_grad: bool = True):
+        self._is_sharded = False
```
`FullyShardedDataParallel` will patch `is_sharded` to FlatParameter. So I initialize it here to make it explicitly.

zhaojuanmao commented:
synced up offline, let's addressed the nits and ran some benchmarks as well

zhaojuanmao reviewed (approved):
looks great!

fegin committed:
[FSDP] Add FlatParameter to track the information of a flat parameter (#69241)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/69241

Implement FlatParameter to track the information of a flat parameter, including the sharding information.

Test Plan: CI

Reviewed By: zhaojuanmao

Differential Revision: D32432503

fbshipit-source-id: 4b3e84f2b4c271ecbea2a3baa9af59c4feb57687
